crates/ruff_linter/src/rules/pylint/rules/unnecessary_assignment.rs

```diff
@@ -0,0 +1,202 @@
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// if test1 := "example":
+///     print(test1)
+/// ```
+///
+/// ## References
+/// - [PEP 572 – Assignment Expressions](https://peps.python.org/pep-0572/)
+/// - [What’s New In Python 3.8 - Assignment Expressions](https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions)
+#[violation]
+pub struct UnnecessaryAssignment {
+    name: Name,
+    assignment: String,
+    parentheses: bool,
+}
+
+impl UnnecessaryAssignment {
+    fn get_fix(&self) -> String {
+        let UnnecessaryAssignment {
+            name,
+            assignment,
+            parentheses,
+        } = self;
+        if *parentheses {
+            format!("({name} := {assignment})")
+        } else {
+            format!("{name} := {assignment}")
+        }
+    }
+}
+
+impl Violation for UnnecessaryAssignment {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use walrus operator `{}`.", self.get_fix())
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some(format!(
+            "Move variable assignment into if statement using walrus operator `{}`.",
+            self.get_fix()
+        ))
+    }
+}
+
+type AssignmentBeforeIfStmt<'a> = (&'a Expr, &'a ExprName, &'a StmtAssign);
+
+/// PLR6103
+pub(crate) fn unnecessary_assignment(checker: &mut Checker, stmt: &StmtIf) {
+    if checker.settings.target_version < PythonVersion::Py38 {
+        return;
+    }
+
+    let if_test = stmt.test.as_ref();
+    let semantic = checker.semantic();
+    let mut errors: Vec<AssignmentBeforeIfStmt> = Vec::new();
+
+    // case - if check (`if test1:`)
+    if let Some(unreferenced_binding) = find_assignment_before_if_stmt(semantic, if_test, if_test) {
+        errors.push(unreferenced_binding);
+    };
+
+    // case - bool operations (`if test1 and test2:`)
+    if let Expr::BoolOp(expr) = if_test {
+        errors.extend(
+            expr.values
+                .iter()
+                .filter_map(|bool_test| {
+                    find_assignment_before_if_stmt(semantic, if_test, bool_test)
+                })
+                .collect::<Vec<AssignmentBeforeIfStmt>>(),
+        );
+    }
+
+    // case - compare (`if test1 is not None:`)
+    if let Expr::Compare(compare) = if_test {
+        if let Some(error) = find_assignment_before_if_stmt(semantic, if_test, &compare.left) {
+            errors.push(error);
+        };
+    }
+
+    // case - elif else clauses (`elif test1:`)
+    errors.extend(
+        stmt.elif_else_clauses
+            .iter()
+            .filter(|elif_else_clause| elif_else_clause.test.is_some())
+            .filter_map(|elif_else_clause| {
+                let elif_check = elif_else_clause.test.as_ref().unwrap();
+                find_assignment_before_if_stmt(semantic, elif_check, elif_check)
+            })
+            .collect::<Vec<AssignmentBeforeIfStmt>>(),
```

Comment:
Changing assignments to walrus operators in `elif else` branches is semantically incorrect if the variable is used afterwards

```python
>>> if True: ...
... elif x :=10: ...
... 
Ellipsis
>>> x
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'x' is not defined
```